### PR TITLE
GDB-5309 Basic authenticate against a secured GDB server

### PR DESCRIPTION
**Pull request by teodossidossev**

### CODE REVIEW

File: src/http/client-config.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    this.pass = pass;
+    this.keepAlive = keepAlive !== undefined ? keepAlive : true;
+  }
+
+  /**
+   * Sets the server's endpoint URL.
+   *
+   * @param {string} endpoint the endpoint URL
+   *
+   * @return {ServerClientConfig} the current config for method chaining
```
Comment (svilenvelikov) [inline src/http/client-config.js:31]: This is the `ClientConfig`, not `ServerClientConfig`

File: test-e2e/tests/auth.spec.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const {ServerClient} = require('graphdb').server;
+const path = require('path');
+const Utils = require('utils');
+const Config = require('config');
+
+describe('Should test auth', () => {
+  let rdfClient = new RDFRepositoryClient(Config.restApiBasicAuthConfig);
+
+  beforeAll((done) => {
+    Utils.createRepo(Config.testRepoPath).then(() => {
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:12]: Can you flatten these promises please?

File: test-e2e/tests/utils.js [hunk 3]
```diff
@@ -53,11 +53,41 @@ function createRepo(path) {
   });
 }
 
+
+function createSecuredRepo(path) {
+  const data = new FormData();
+  data.append("config", fs.createReadStream(path));
```
Comment (svilenvelikov) [inline test-e2e/tests/utils.js:59]: Single quotes

File: test-e2e/tests/utils.js [hunk 4]
```diff
@@ -53,11 +53,41 @@ function createRepo(path) {
...
     url: `${Config.serverAddress}/rest/repositories/${name}`
   });
 }
 
-module.exports = {loadFile, readStream, getReadStream, createRepo, deleteRepo};
+function toggleSecurity(enable) {
+    return axios({
+      method: 'post',
+      headers: { "Content-Type": "application/json", "Connection": "keep-alive", "Accept": "*/*"},
+      url: `http://localhost:7200/rest/security?useSecurity=${enable}`,
```
Comment (svilenvelikov) [inline test-e2e/tests/utils.js:83]: Use the serverAddress config here

File: test-e2e/tests/utils.js [hunk 5]
```diff
@@ -53,11 +53,41 @@ function createRepo(path) {
...
+      data: `${enable}`,
```
Comment (svilenvelikov) [inline test-e2e/tests/utils.js:84]: make it just `data: enable,`

Comment (svilenvelikov) [reply test-e2e/tests/utils.js:84]: or is it a string `true`|`false`?

File: test-e2e/tests/auth.spec.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    });
+  });
+
+  afterAll(() => {
+    return Utils.toggleSecurity(false).then(() => {
+      return Utils.deleteRepo('Test_repo');
+    });
+  });
+
+  test('Should list namespces', () => {
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:28]: `namespces` -> `namespaces`

File: test-e2e/tests/auth.spec.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+const {RDFRepositoryClient} = require('graphdb').repository;
+const {RDFMimeType} = require('graphdb').http;
+const {ServerClient} = require('graphdb').server;
+const path = require('path');
+const Utils = require('utils');
+const Config = require('config');
+
+describe('Should test auth', () => {
+  let rdfClient = new RDFRepositoryClient(Config.restApiBasicAuthConfig);
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:9]: const

File: test-e2e/tests/auth.spec.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+  beforeAll((done) => {
+    Utils.createRepo(Config.testRepoPath).then(() => {
+      let wineRdf = path.resolve(__dirname, './data/wine.rdf');
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:13]: const

File: test-e2e/tests/auth.spec.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  test('Should delete all namespaces', () => {
+    return rdfClient.deleteNamespaces().then(() => {
+      return rdfClient.getNamespaces();
+    }).then((resp) => {
+      expect(resp.length).toBe(0)
+    });
+  });
+
+
+  test('Should delete secured repo via server client',  () => {
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:53]: Did you configure the eslint for the project? There are couple of issues here.

File: test-e2e/tests/auth.spec.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    Utils.createSecuredRepo(Config.testRepo2Path);
+    const serverClient = new ServerClient(Config.serverBasicAuthConfig);
+    serverClient.deleteRepository('Test_repo_2').then(()=> {
+      return serverClient.hasRepository('Test_repo_2').then((resp) => {
```
Comment (svilenvelikov) [inline test-e2e/tests/auth.spec.js:57]: Don't nest promises, please

File: src/repository/repository-client-config.js [hunk 11]
```diff
@@ -6,7 +8,7 @@
  * @author Mihail Radkov
  * @author Svilen Velikov
  */
-class RepositoryClientConfig {
+class RepositoryClientConfig extends ClientConfig {
```
Comment (svilenvelikov) [inline src/repository/repository-client-config.js:11]: good that you've extracted this basic config

File: src/http/client-config.js (hunk 1 above)
Comment (svilenvelikov) [reply src/http/client-config.js:31]: Keep in mind that the methods here now can return either ServerClientConfig or RepositoryClientConfig according to what's the concrete instance type, so please fix the docs in this class.

File: README.md [hunk 12]
```diff
@@ -461,7 +465,19 @@ const config = new RepositoryClientConfig()
...
+JWT is serialized as “Authorization: GDB <token>” header in every request, so it is vulnerable to a man-in-the-middle attack. Everyone who intercepts the JWT can reuse the session. To prevent this, we recommend to always enable encryption in transit.
+
+#### Basic Authentication¶
+Instead of using JWT, users can access secured GraphDB by passing valid base-64 encoded username:password combinations as a header.
+In case Basic authentication will be used, then the headers in the `ServerClientConfig` and `RepositoryClientConfig` must be configured to send the `username` and `password` which to be used for the authentication. From this moment on, with every API call is sent also an `authorization` header with the encoded credentials as value.
+```javascript
+config.setBasicAuthentication('admin:root');
+```
+
+##### Note:
```
Comment (svilenvelikov) [inline README.md:479]: You can also put this section in a blockquote and maybe bold s that it stands out.

File: src/http/client-config.js (hunk 1 above)
Comment (teodossidossev) [reply src/http/client-config.js:32]: done

File: test-e2e/tests/auth.spec.js (hunk 2 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js:12]: done

File: test-e2e/tests/utils.js (hunk 3 above)
Comment (teodossidossev) [reply test-e2e/tests/utils.js]: done

File: test-e2e/tests/utils.js (hunk 4 above)
Comment (teodossidossev) [reply test-e2e/tests/utils.js]: done

File: test-e2e/tests/utils.js (hunk 5 above)
Comment (teodossidossev) [reply test-e2e/tests/utils.js:84]: it's a string

File: test-e2e/tests/auth.spec.js (hunk 7 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js]: done

File: test-e2e/tests/auth.spec.js (hunk 8 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js]: done

File: test-e2e/tests/auth.spec.js (hunk 9 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js]: fixed

File: test-e2e/tests/auth.spec.js (hunk 10 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js]: fixed

File: README.md (hunk 12 above)
Comment (teodossidossev) [reply README.md]: done

File: test-e2e/tests/auth.spec.js (hunk 6 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js]: moved

File: test-e2e/tests/auth.spec.js (hunk 2 above)
Comment (svilenvelikov) [reply test-e2e/tests/auth.spec.js:12]: No, it's not. It should be like: 
```javascript
    Utils.createRepo(Config.testRepoPath).then(() => {
      const wineRdf = path.resolve(__dirname, './data/wine.rdf');
      return rdfClient.addFile(wineRdf, RDFMimeType.RDF_XML, null, null);
    }).then(() => {
      return Utils.toggleSecurity(true);
    }).then(() => {
      done();
    });
```

File: test-e2e/tests/namespaces.spec.js [hunk 13]
```diff
@@ -41,3 +40,51 @@ describe('Should test namespaces', () => {
     });
   });
 });
+
+describe('Should test namespaces in secured environment', () => {
+  const rdfSecuredClient =
+    new RDFRepositoryClient(Config.restApiBasicAuthConfig);
+
+  beforeAll((done) => {
```
Comment (svilenvelikov) [inline test-e2e/tests/namespaces.spec.js:48]: see the other spec and make these promises flat too

File: test-e2e/tests/namespaces.spec.js [hunk 14]
```diff
@@ -41,3 +40,51 @@ describe('Should test namespaces', () => {
...
+
+  test('Should list namespces', () => {
+    rdfSecuredClient.getNamespaces().then(() => {
+      return rdfSecuredClient.getSize();
+    }).then((response) => {
+      expect(response).toBe(1839);
+    });
+  });
+
+  test('Should delete all statements', () => {
```
Comment (svilenvelikov) [inline test-e2e/tests/namespaces.spec.js:75]: But this is a test for statements, not for namespaces. Move it in the other spec

Comment (teodossidossev) [reply test-e2e/tests/namespaces.spec.js]: done

File: test-e2e/tests/namespaces.spec.js (hunk 13 above)
Comment (teodossidossev) [reply test-e2e/tests/namespaces.spec.js:48]: done

File: test-e2e/tests/auth.spec.js (hunk 2 above)
Comment (teodossidossev) [reply test-e2e/tests/auth.spec.js:12]: my bad. Done

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**